[WIP] Swift 3.0 Compatibility

norio-nomura opened the pull request:
- Uses wrapper that converting from Swift 3.0 API to Swift 2.2 API.
- Includes #261
- Uses swift-DEVELOPMENT-SNAPSHOT-2016-03-24-a
- Passes `swift test` on OS X and Linux
- [x] Update test script for Linux

norio-nomura commented:
It seems updating test script for Linux is not needed.

briancroom commented:
Wow, this looks really great @norio-nomura! I'm especially impressed that you were able to achieve such good 2.2-3.0 compatibility. Could you rebase this when you get a chance?

norio-nomura commented:
Rebased.

norio-nomura commented on Sources/Nimble/Swift3to22.swift:
```diff
@@ -0,0 +1,157 @@
+        }
+    }
+
+    public protocol Sequence: SequenceType {
```
For supporting uses `associatedtype` constraints on some functions, `Sequence` is defined as `protocol` inheriting `SequenceType`.
That would require some `SequenceType` types defined on consumer to conform as:
```swift
#if !swift(>=3)
extension MyCustomSequence: Nimble.Sequence {}
#endif
```

briancroom commented:
Hi @norio-nomura, I'm sorry that there's been so little activity around here for so long.

I'm inclined to say that we're best off introducing a `swift3` branch for this for now, since the development snapshots are such a fast-moving target at this point and I feel like there's going to be a good bit of churn needed here to keep up. I would tentatively suggest that we could put this on master and do lightweight maintenance of a `swift2.2` branch around WWDC-time, when official Swift 3 beta releases will presumably start to be pushed out.

Thoughts @norio-nomura @Quick/contributors?

norio-nomura commented:
This PR has already been incompatible with `swift-DEVELOPMENT-SNAPSHOT-2016-04-12` because of [SE-0046 Establish consistent label behavior across all parameters including first labels](https://github.com/apple/swift-evolution/blob/master/proposals/0046-first-label.md).
Almost of functions and methods need to be enclosed with `#if swift(>=3)` for supporting both of swift 2.2 and 3. e.g. https://github.com/antitypical/Result/pull/149/files

norio-nomura commented:
I can't say which is better whether supporting swift-2.2 and 3 simultaneously or branching them.

modocache commented:
I'd like to reiterate my thought that XCTest should export both APIs until Swift 3 is officially released: [SR-1215](https://bugs.swift.org/browse/SR-1215). We're doing very little to make things smooth for early adopters of corelibs-xctest, I think we could do better in that respect.

briancroom commented:
In light of the comments on this PR, [this discussion](https://bugs.swift.org/browse/SR-1215), and the evidence of the complexity of supporting 2.2 and 3 simultaneously as [linked to above](https://github.com/antitypical/Result/pull/149/files), I'm pretty convinced at this point that branching is required here. Additionally, I feel that the `swift3` branch will need to track the development snapshots pretty closely for the next while, as breaking changes continue to roll in. Given all that, I don't think there is much value in keeping the `Swift3to22.swift` shim in this PR.

@norio-nomura how would you feel about updating this to work with `swift-DEVELOPMENT-SNAPSHOT-2016-04-25`? I think we can use this as a start of a `swift3` branch at that point.

modocache commented:
Makes sense to me! Theoretically it's possible to have CI build the Swift 3 branch by compiling Swift from source, then passing `SWIFT_EXEC` to `xcodebuild`. That seems like overkill, though--should we just go without CI for Swift 3?

norio-nomura commented:
I'd happy to update `nn-swift-3-compatibility` to work with `swift-DEVELOPMENT-SNAPSHOT-2016-04-25` and drop supporting swift-2.2. I hope I would do that before next snapshot coming.

briancroom commented:
Hmm, I wonder how feasible it would be to download and install a toolchain package until Travis starts getting Xcode betas that include a Swift 3 toolchain? Even if we decide not to do that, the Linux CI should still Just Work here, as it is already downloading and installing the specified dev snapshot.

ashfurrow commented:
If Travis is difficult with OS X builds, I think removing them and keeping Linux for now would be a reasonable compromise.

norio-nomura commented:
We can use Swift 3 with Xcode 7.3 on Travis-CI as following: https://github.com/norio-nomura/SWXMLHash/blob/nn-swift22-and-swift3-compatibility/.travis.yml#L11-L18

briancroom commented:
That looks worth trying to me!

norio-nomura commented:
Rebased and updated to `swift-DEVELOPMENT-SNAPSHOT-2016-05-03-a`.
- [ ] podspec fails since `swift-DEVELOPMENT-SNAPSHOT-2016-04-12-a`
- [ ] `BeCloseToTest` fails on linux since `swift-DEVELOPMENT-SNAPSHOT-2016-05-03-a`

norio-nomura commented:
Updated to `swift-DEVELOPMENT-SNAPSHOT-2016-05-09-a`.
Now `test podspec` is only failing test on Travis-CI.
Do we need to fix that for creating `swift3` branch from this PR?

norio-nomura commented:
By using cocoapods-1.0.0, `test podspec` passed on Travis-CI. 😃

liscio commented:
Just FYI I have opened a PR against Norio's fork here: https://github.com/norio-nomura/Nimble/pull/1

It now builds & passes the tests for both the iOS & Mac targets.

norio-nomura commented:
Now, I updated to swift-3.0-preview-1 and all tests are passed on my Mac.
On Travis-CI:
- test randomly fails
- podspec is disabled because of https://github.com/CocoaPods/CocoaPods/issues/5519

NachoSoto commented:
I pushed the necessary changes for beta 2 here: https://github.com/norio-nomura/Nimble/pull/2/files :)

norio-nomura commented:
I updated, but travis is not yet.

andersio commented:
Lower deployment target to 10.9: https://github.com/norio-nomura/Nimble/pull/3
Please have a look.

:)

norio-nomura commented:
I updated, but travis is not yet.
On linux, it failed with https://github.com/kylef/swiftenv/pull/47

norio-nomura commented:
Tests passed on `swift-3.0-PREVIEW-2`

mokagio commented:
FYI, I did some work for Xcode 8 Beta 4 https://github.com/norio-nomura/Nimble/pull/5. There are two issues that I haven't yet been able to work around, I'd appreciate feedback on them.

norio-nomura commented:
Now this PR is compatible with swift-3.0-PREVIEW-5. Thanks @mokagio!

ikesyo commented:
WIP Xcode 8 beta 6 support is here: https://github.com/norio-nomura/Nimble/pull/6. Feel free to take over it!

raphaelcruzeiro commented:
Did you have success using this branch of Nimble to tests actual projects? I'm trying to convert my tests to use this version of Nimble with Swift 3 and I have errors everywhere I use `beCloseTo`. Namely, `Generic parameter 'T' could not be inferred` and `Ambiguous reference to member 'beCloseTo(_:within:)'`.  My guess is that Nimble is not working well with `CGFloat`.

ikesyo commented:
ReactiveCocoa did: https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3145

norio-nomura commented:
@raphaelcruzeiro `CGFloat` was implicitly converted to `NSNumber` on Swift 2.2.
[SE-0072](https://github.com/apple/swift-evolution/blob/master/proposals/0072-eliminate-implicit-bridging-conversions.md) removes those conversions.
Casting your `CGFloat` to `NSNumber` will help you.
Of course PR will be welcomed. 😉

norio-nomura commented:
Maybe making `CGFloat` conforming to `NMBDoubleConvertible` will be better resolution.

raphaelcruzeiro commented:
@norio-nomura Agreed. I'm gonna start working on that then ;)

norio-nomura commented:
Now `CGFloat` conforms to `NMBDoubleConvertible` by https://github.com/norio-nomura/Nimble/pull/7
Thank you @raphaelcruzeiro!

pcantrell commented:
@raphaelcruzeiro: Siesta has been using this fork successfully (and the corresponding Quick fork as well). On the `beCloseTo` question, I’ve had success using `≈` and `±`:

https://github.com/bustoutsolutions/siesta/blob/swift3/Tests/ProgressSpec.swift#L112

I was even able to remove a `Double(…)` coercion that Swift 2 required.

Thanks, @norio-nomura, for keeping these forks going!

raphaelcruzeiro commented:
Now that `CGFloat` conforms to `NMBDoubleConvertible` it should work fine for all the matchers.

And yeah! Thanks to @norio-nomura for taking the step forward and getting this branch up and working!

ikesyo commented:
I'd like to officially manage Swift 3 migration in this repo and I've created `swift-3.0` branch. @norio-nomura Could you change the base branch of this PR to that branch? Once that is done, I would merge this and submit a new PR from the branch to `master`. Then, subsequent PRs for Swift 3 should be targeted on the branch.

norio-nomura commented:
@ikesyo Thanks! I have changed base branch to `swift-3.0`. 🙏

ikesyo commented:
Thank you @norio-nomura and all who contributed to Swift 3 support on this PR! 🎉
